src/Pact/Types/Hash.hs

```diff
@@ -1,49 +1,186 @@
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE DeriveGeneric #-}
 {-# LANGUAGE CPP #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
 module Pact.Types.Hash
   (
   -- Bytestring hashing
-    hash
+    Hash(..)
+  , hash
+  , hashToText
   , verifyHash
   , initialHash
+  , hashLength
+  , TypedHash(..)
+  , toUntypedHash, fromUntypedHash
+  , HashAlgo(..)
+  , pactHash, pactInitialHash
   ) where
 
 
-import Prelude hiding (null)
+import Control.DeepSeq
 import Data.ByteString (ByteString)
-
+import Data.Hashable (Hashable(hashWithSalt))
+import Data.Serialize (Serialize(..))
 import Pact.Types.Pretty
 import Pact.Types.Util
-
+import Data.Aeson
+import GHC.Generics
+import Data.Text (Text)
+import Data.Text.Encoding (decodeUtf8)
+import Data.Reflection
+import Data.Proxy
 
 #if !defined(ghcjs_HOST_OS)
 
 import qualified Data.ByteArray as ByteArray
 import qualified Crypto.Hash as Crypto
 
-hash :: ByteString -> Hash
-hash = Hash . ByteArray.convert . Crypto.hashWith Crypto.Blake2b_512
-{-# INLINE hash #-}
-
 #else
 
 import Crypto.Hash.Blake2Native
 
-hash :: ByteString -> Hash
-hash bs = case blake2b hashLengthAsBS mempty bs of
-  Left _ -> error "hashing failed"
-  Right h -> Hash h
-
 #endif
 
 
-verifyHash :: Hash -> ByteString -> Either String Hash
-verifyHash h b = if hash b == h
-  then Right h
+
+-- | Untyped hash value, encoded with unpadded base64url.
+-- Within Pact these are blake2b_256 but unvalidated as such,
+-- so other hash values are kosher (such as an ETH sha256, etc).
+newtype Hash = Hash { unHash :: ByteString }
+  deriving (Eq, Ord, Generic, Hashable, Serialize)
+
+instance Show Hash where
+  show (Hash h) = show $ encodeBase64UrlUnpadded h
+
+instance Pretty Hash where
+  pretty = pretty . asString
+
+instance AsString Hash where
+  asString (Hash h) = decodeUtf8 (encodeBase64UrlUnpadded h)
+
+instance NFData Hash
+
+instance ToJSON Hash where
+  toJSON = String . hashToText
+
+instance FromJSON Hash where
+  parseJSON = withText "Hash" parseText
+  {-# INLINE parseJSON #-}
+
+instance ParseText Hash where
+  parseText s = Hash <$> parseB64UrlUnpaddedText s
+  {-# INLINE parseText #-}
+
+
+hashToText :: Hash -> Text
+hashToText (Hash h) = toB64UrlUnpaddedText h
+
+
+-- | All supported hashes in Pact (although not necessarily GHCJS pact).
+data HashAlgo =
+  Blake2b_256 |
+  SHA3_256
```

Comment:
If we're only allowing `SHA3` here for Ethereum payload hashes, I don't think its necessary (see previous comment). 
I use `hashTx` with `SHA3` in one of the Scheme tests to check a failure test, but I can just call Cryptonite's hash function directly in this instance.

Reply:
See other comment, this is just to ensure we don't have only one case, as we will need to support multiple algos within Pact (e.g. Chainweb uses SHA3_256, Bitcoin uses SHA2, etc)